text/0000-light-dom.md

```diff
@@ -0,0 +1,218 @@
+  }
+}
+
+// doesn't use Shadow DOM
+class MyLightElement extends HTMLElement {
+  connectedCallback() {
+    this.innerHTML = "<p>My element that does not use Shadow DOM";
+  }
+}
+```
+
+In LWC as well, the developer will retain the choice: they can choose to inherit from `LightningElement` to select Shadow DOM (which already exists) or inherit from a new class, `MacroElement`, to opt into Light DOM.
+
+```js
+import { MacroElement } from "lwc";
+
+export default class MyLightComponent extends MacroElement {
+  renderedCallback() {
+    this.querySelector("child-element"); // note the absence of this.template
+  }
+}
+```
+
+`MacroElement` is a new class that will be exported from the `lwc` module. Its API and functionality remain identical to `LightningElement` with the primary difference of not using Shadow DOM.
+
+### Component features when using Light DOM
+
+Some of the LWC component capabilities are directly inherited from Shadow DOM, or emulated by the synthetic-shadow. Despite the use of Light DOM, we’d like to keep these features available, even if their behavior is slightly adapted to the Light DOM:
+
+#### Slots
+
+As we mentioned before, the component composition model in LWC is provided by slots. Light DOM will provide the same mental model for developers building Light DOM components.
+
+In Light DOM, `<slot>` will denote the place where the slotted component will be attached. The `<slot>` element itself won't be rendered. The slotted content (or the fallback content) will be flattened to the parent element at runtime.
+
+Since the `<slot>` element itself isn't rendered, adding attributes or event listeners to the `<slot>` element in the template will throw a compiler error.
+
+#### Styles
+
+Until now styles used in LWC components where scoped to the component thanks to shadow DOM (or synthetic shadow DOM) style scoping. In the light DOM, component styles are naturally leaks out of the component, LWC doesn't do any style scoping out of the box. Developers are in charge of making sure to write specific enough selector to target the component content.
+
+To support the cases where a shadow DOM element composes a light element, light DOM styles are required to be injected to the closest root node. For a given light DOM elements if all the ancestor components are also a light DOM components, the component style sheet will be injected in the document `<head>`. Otherwise if any of the ancestors is a shadow DOM components the style has to be injected in the closest shadow root. Upon insertion of a light DOM element does the following steps:
+
+- look up for the closest root node (`Node.prototype.getRootNode()`)
+- insert the stylesheet if not already present:
+  - if the root node is the HTML document, the style sheet is inserted as a `<style>` element in the `<head>`.
+  - if the root node is a shadow root, the stylesheet is inserted as a `<style>` element as the first shadow root child.
```

Comment:
This would be a good opportunity to align with [constructable stylesheets](https://developers.google.com/web/updates/2019/02/constructable-stylesheets), which is already available in Chrome and fairly easy to polyfill/workaround in other browsers.

The main benefit of this API is that, if some custom element is repeated several times in the document, then this is an explicit signal to the browser that it can optimize that stylesheet's use across multiple elements. Otherwise if you have 1,000 shadow DOM elements each with an identical light DOM element inside, we would have to create 1,000 identical `<style>`s and inject them into the `shadowRoot`. (Presumably browsers could optimize this based on a checksum of the `<style>` content or something, but for whatever reason the Chrome devs seem to prefer constructable stylesheets as a more explicit signal.)

I would propose using constructable stylesheets in Chrome, and the above steps for other browsers. In the case of Chrome, it's just a matter of doing the exact same thing (targeting either the `head` or the nearest `shadowRoot`) but instead of inserting a `<style>` element, we append to [`documentOrShadowRoot.adoptedStyleSheets`](https://wicg.github.io/construct-stylesheets/#proposed-solution).

(N.B. this would require disallowing `@import`, but I assume we already do this? If not, we would have to.)

Off-topic: we could already switch to using this today for synthetic shadow DOM, AFAICT. /cc @ravijayaramappa

Reply:
@pmdartus Do these `@import` statements actually end up in the final CSS, or is it just an authoring concern? If it ends up in the final CSS, then using constructable stylesheets gets much more complicated, because we would need to use the async API rather than the sync API.

Reply:
No, the imports are stripped out. Those imports are resolved by the application container. This is why the [`TemplateStylesheetFactories`](https://github.com/salesforce/lwc/blob/15ff9bb98ec324dc68d40c96e6d37de69d9f1145/packages/%40lwc/engine-core/src/framework/stylesheet.ts#L25-L30) is recursive as it might contain other stylesheets.

Reply:
Constructable stylesheets may be a perf optimization we can do later. Resolving for now.